compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala

```diff
@@ -2042,24 +2042,44 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
     }}
     val argVals = argVals0.reverse
     val argRefs = argRefs0.reverse
-    def rec(fn: Tree): Tree = fn match {
+    def rec(fn: Tree, topAscription: Option[TypeTree]): Tree = fn match {
+      case Typed(expr, tpt) =>
+        // we need to retain any type ascriptions we see and:
+        // a) if we succeed, ascribe the result type of the ascription to the inlined body
+        // b) if we fail, re-ascribe the same type to whatever it was we couldn't inline
+        // note: if you see many nested ascriptions, keep the top one as that's what the enclosing expression expects
+        rec(expr, topAscription.orElse(Some(tpt)))
       case Inlined(call, bindings, expansion) =>
         // this case must go before closureDef to avoid dropping the inline node
-        cpy.Inlined(fn)(call, bindings, rec(expansion))
+        cpy.Inlined(fn)(call, bindings, rec(expansion, topAscription))
       case closureDef(ddef) =>
         val paramSyms = ddef.vparamss.head.map(param => param.symbol)
         val paramToVals = paramSyms.zip(argRefs).toMap
-        new TreeTypeMap(
+        val result = new TreeTypeMap(
           oldOwners = ddef.symbol :: Nil,
           newOwners = ctx.owner :: Nil,
           treeMap = tree => paramToVals.get(tree.symbol).map(_.withSpan(tree.span)).getOrElse(tree)
         ).transform(ddef.rhs)
+        topAscription match {
+          case Some(tpt) =>
+            // we assume the ascribed type has an apply that has a MethodType with a single param list (there should be no polys)
+            val methodType = tpt.tpe.member(nme.apply).info.asInstanceOf[MethodType]
```

Comment:
@fhackett won't this crash if the ascribed type has a `val apply` member that's not a method? I think this could happen if the term passed as `fn` is not really a function, or if the `apply` is overloaded. Something like:

```scala
scala> object test extends (String => String) { val apply = (x:Int) => x; def apply(x: String) = x }
// defined object test

scala> test.apply("abc")
val res0: String = abc

scala> test.apply
val res1: Int => Int = test$$$Lambda$1268/578362229@6f3bd37f
```

Reply:
This case is only reached if the prefix is an explicit closure.

Reply:
Right, I had missed the `closureDef` pattern. But it seems even a `Closure` can have a type that contains an overloaded `apply` member: indeed it [can be any SAM type](https://github.com/lampepfl/dotty/blob/a29178ce4ac848e773423f57f45186373f410970/compiler/src/dotty/tools/dotc/ast/Trees.scala#L508-L510), no?

Reply:
Maybe, I will have to check if we can get one of those here. @LPTK did you have a specific example in mind?

Reply:
Something like `betaReduce('{ (x => x) : Test }, ...)` where we have:

```scala
abstract class Test extends (String => String) { val apply = 123 }.
```

Reply:
I tried it, and we get a class cast exception.
I can figure out a fix - should I push a second commit here or open a second issue+PR?

Reply:
Actually, I can do one "better". See this fun edge case:
```
abstract class Test extends (Int => Int) {
  def impl(i: Int): Int
  def apply(i: Int): Int = ???
}
```

Reply:
This is getting more involved than just a quick fix, so I'll open a fresh issue.

Reply:
Continued on #8290